---
title: "Introduction to GenomicBreaks"
author: 
 - "Charlotte West"
 - "Charles Plessy"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Introduction to GenomicBreaks}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_knit$set(verbose = TRUE)
```

# Introduction

The _GenomicBreaks_ R package contains tools for visualising and analysing
characteristics associated with _breakpoints_ between pairs of _genomes_.

## Load pacakges

Core packages that provide functions we use a lot.

```{r loadlib, echo=T, results='hide'}
suppressPackageStartupMessages({
  library('GenomicBreaks')
  library('BSgenome')
  library('GenomicFeatures')
  library('GenomicRanges')
  library('ggplot2')
})
```

We will use yeast data as an example, so let's load the `Scerevisiae` BSgenome
package.

```{r load_BSgenome_packages}
library("BSgenome.Scerevisiae.UCSC.sacCer3")
```

## Example data

This package contains an example data file representing the alignment of the
_Saccharomyces paradoxus_ genome version `ASM207905v1` to the `SacCer3` reference genome of
_S. cerevisiae_.  An alignment file was produced with a Nextflow
[pairwise alignment pipeline](https://github.com/oist/plessy_pairwiseGenomeComparison)
and converted to [GFF3](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md)
format with the `maf-convert gff -J 1000000` command of the
[LAST](https://gitlab.com/mcfrith/last/-/blob/main/doc/maf-convert.rst) package.

```{r locate_example_data}
exdata <- system.file("extdata/SacCer3__SacPar.gff3.gz", package = "GenomicBreaks")
rtracklayer::import(exdata)
```

# GBreaks objects

Pairwise genome alignments are loaded in `GBreaks()` objects that wrap the
`GRanges` class.  By convention, we write that the _query_ genome (the one that
was provided as a FASTA file) is aligned to the _target_ genome (the one that
was indexed by the aligner).  The _target_ genome (on the left) is the main part
of the object, and the _query_ genome information (right) is contained in the
metadata columns (`mcols`) of the structure.  Information on contig or scaffold
length is optionaly loaded from `BSgenome` objects, from which `seqinfo` data
is extracted and passed to the `GRanges`.  By convention, strand information
of the alignment is held by the _target_ `GRanges`, and the _query_ `GRanges`
are strandless.  The objects are sorted by `seqname` first.

```{r GRanges object}
gb <- load_genomic_breaks(exdata, Scerevisiae)
gb
gb$query
```

## Maniuplation of GBreaks objects with plyranges

With the _plyranges_ package it is easy to manipulate `GBreaks` objects, in
particular when filtering on features of the _query_ genome.

```{r plyranges_example}
# Subset for chrI on the target genome.
gb |> plyranges::filter(seqnames == "chrI")

# Subset for NC_047487.1 on the query genome
gb |> plyranges::filter(seqnames(query) == "NC_047487.1")

# Add or modify columns on the fly
gb |> plyranges::mutate("Width" = width) |> head(3)
gb |> plyranges::mutate("score" = width) |> head(3)

# Etc…
```

See the [vignette of _plyranges_](https://bioconductor.org/packages/release/bioc/vignettes/plyranges/inst/doc/an-introduction.html) for more details.

## Breakpoints and alignment stops

Let us define an **alignment stop** to be a position defined in either the
_target_ or _query_ genome, where an alignment begins or ends.  Looking at the
first line in the `gb` object above, `ChrI` positions 5860 and 10010 of the
_target_ would be defined as alignment stops.  Correspondingly, we have
positions 8723 and 12176 in `NC_047487.1` of the _query_ genome as alignment
stops. 

The function `get_bps()` takes in a `GBreaks` pairwise alignment or a simple
`GRanges` object, and produces a `GRanges` object of it's associated alignment
stops.  It has `direction`, `stranded` and `sorted` options to refind its
output. 

```{r get bps}
get_bps(gb) |> head(3)
get_bps(gb, direction = "left") |> head(3)
get_bps(gb,                      stranded = TRUE) |> head(3)
get_bps(gb,                      stranded = TRUE, sorted = FALSE) |> head(3)
get_bps(gb, direction = "right", stranded = TRUE) |> head(3)
get_bps(gb$query, sorted = FALSE) |> head(3)
get_bps(gb$query) |> head(3)
```

Now, let us define a **genomic breakpoint** (or simply breakpoint) to be a
genomic structural mutation, arising from breakage and repair of the chromosome.
Such structural events include insertion, deletion, inversion and translocation,
and often arise during recombination. 

This package seeks to determine, of the group of alignment stops, which are
likely genomic breakpoints and analyse them.  Thus, filtering out alignment
stops that are a result of bioinformatic errors.

## Arms

If the positions of the centromeres are known, an annotation file can be
crafted to indicate the coordinates of the short and long arms, and the
function `flagLongShort()` can be used to transfer this annotation to the
`GBreaks` object.

```{r}
annot <- GRanges(c("chrI:1-151465", "chrI:151582-230218"))
annot$Type <- c("short", "long")
flagLongShort(gb, annot)
```

## Measures of distance

### Synteny index

Ad-hoc index measuring to what extent a scaffold of the _target_ genome is
mostly aligned to a single scaffold in the _query_ genome.   See
`?synteny_index` for details.  One limitation to the use of this
index is that it requires that at least the _query_ genome is a complete
chromosome assembly.

```{r}
synteny_index(gb)
synteny_index(swap(gb))
```

### Correlation

Ad-hoc index measuring the correlation of the coordinates of the syntenic
alignments in scaffolds of a _target_ genome and their best match in the
_query_ genome.  See `?correlation_index` for details.  This
index is more robust to the presence of uncollapsed haplotypes in the _query_
genome.

```{r}
correlation_index(gb)
correlation_index(swap(gb))
```

### Gene Order Conservation

The `GOC()` function calcluates the Gene Order Conservation number defined by
Rocha (2003, <https://doi.org/10.1016/j.tig.2003.09.011>) as “_the average
number of orthologues for which the consecutive orthologue co-occurs close by in
the other genome. It varies between 0 (no co-occurrence) and 1 (complete gene
order conservation)_”.

_Note that calculating GOC on whole-genome alignments is not expected
to produce meaningful results, and the example below is just to show that the
computation is possible.  This function is more useful when comparing the
position of orthologues, represented in a `GBreaks` object._

```{r}
GOC(gb)
GOC(swap(gb))
```

## Visualising alignments

We can use the function `scaf_align_plot` to see the alignment of query
scaffolds/chromosomes onto a specific target scaffold/chromosome of interest.

```{r echo=T, message=F, warning=F}
scaf_align_plot(gb, scaf = "chrI")
```

Here we have plotted the alignments on the Okinawa genome (y-axis) onto the Osaka genome, scaffold 2 (x-axis). The coverage of the alignment is the bar along the top. 

We can plot alignments for two of the target scaffolds side-by-side, using `align_scaf_plot-2`. This can be used in aid of superscaffolding. 

```{r echo=T, message=F, warning=F}
align_scaf_plot_2(gb, scaf = c("chrI","chrII"))
```

As you can see, scaffold 2 of the query genome aligns to the majority of both target scaffolds. This could be used as evidence to support the superscaffolding of these two scaffolds - in an effort toward chromosomal assembly. 

## Coalescing alignments

Large syntenic regions can often appear cluttered with alignment breaks,
spanning just a few basepairs.  They are either an artefact (for instance in
case of incomplete purge of haplotypes) and or true breakpoint.  The algorithm
in `coalesce_contigs` is used to produce a new GRanges object with fewer
alignment breaks by coalescing alignments separated by short (user specified)
distances.  This distance need be agreeable in both the target and query genome
in order for the coalscion to happen. 

For example, coalescing gaps of less than 500 basepairs in the `gb` alignment:

```{r coalescing algorithm}
coa <- coalesce_contigs(gb)
length(gb)
length(coa)
```

The resulting `GBreaks` object has far fewer alignments and therefore far fewer
alignment stops. The algorithm is an initial step in alignment stop filtering,
with the goal of a reduced number of alignment stops that have a high
probability of being breakpoints.

## Genome plots with genoPlotR

To visualise synteny it is clearer to plot from the coalseced objects.

```{r test_genoPlotR}
plotApairOfChrs(gb,  "chrI")
plotApairOfChrs(coa, "chrI")
```

## Trivial translocations

### Definitions

Remember that we are working with one-to-one alignments.  Therefore deletions,
insertions and as a consequence copy number variations, are not in the aligned
regions.  Nevertheless, we can detect translocations from one chromosomal
region to another.  We define them by alignment pairs that interrupt the
colinearity between two other pairs, and that are not inversions.

### Representation

The translocation below:

```
┌──────────────┬──────────────┬──────────────┐
│ chrA:101-200 │ chrA:201-300 │ chrA:301-400 │ (Target genome)
└──────────────┴──────────────┴──────────────┘
      +               +             +         (Alignment direction)
┌──────────────┬──────────────┬──────────────┐
│ chrB:101-200 │ chrC:201-300 │ chrB:301-400 │ (Query genome)
└──────────────┴──────────────┴──────────────┘
```

Is represented as:

```{r ideal_insertions}
exampleTranslocation
plotApairOfChrs(exampleTranslocation)
```

See `vignette("StructuralVariants", package = "GenomicBreaks")` for other
examples of translocations.

Note also that *without an outgroup it is not possible to know in which genome
the move happened*.

### Detection

```{r detect_trivial_insertions}
flagTranslocations(exampleTranslocation)
showTranslocations(flagTranslocations(coa))
plotApairOfChrs(coa, chrT = "chrI", xlim = gb2xlim(coa[2:4]))
```

## Trivial inversions

See `vignette("StructuralVariants", package = "GenomicBreaks")` for definitions.

### Number of trivial inversions

More inversions are found after coalescing colinear blocks because of situations
where `+ - +` was `+ - - +` before collapsing. 

```{r detect_trivial_invertions}
flagInversions(coa)
showInversions(flagInversions(coa))
showInversions(flagInversions(coa)) |> plyranges::slice(1:3)   |> plotApairOfChrs()
showInversions(flagInversions(coa)) |> plyranges::slice(4:9)   |> plotApairOfChrs()
showInversions(flagInversions(coa)) |> plyranges::slice(10:12) |> plotApairOfChrs()
```

### Isolate the left-side gaps in inversions

```{r search_for_inversions}
# lgaps <- function(reduced_gr) {
#   # Flag inversions.
#   #  WHAT FOLLOWS ASSUMES THAT THE FLAG IS ON THE LEFT-SIDE BLOCK OF THE TRIPLE
#   lgaps <- flagInversions(reduced_gr)
#   # Extract inversions
#   invs <- lgaps[lgaps$inv]
#   # Extend inversion of 1 bp so that they overlap with their neighbor gap
#   invs <- shift(invs, 1)
#   # Remove strand information
#   strand(lgaps) <- "*"
#   # Then extract gap positions ignoring strand
#   lgaps <- gaps(lgaps)
#   # Remove bogus gaps on + and - strands
#   lgaps <- lgaps[strand(lgaps) == "*"] 
#   # Return the gaps overlapping with the flagged inversions
#   subsetByOverlaps(lgaps, invs)
# }
# 
# # Using coa and not gbs for reason on better detection explained above.
# inv.lgaps <- sapply(coa, lgaps) |> SimpleList()
# 
# data.frame(width=width(inv.lgaps$Oki_Osa)) |>
#   ggplot(aes(width)) + geom_histogram() +  scale_x_log10()
# 
# inv.lgaps.Seq <- SimpleList()
# 
# inv.lgaps.Seq$Oki_Osa <- BSgenome::getSeq(OKI2018_I69,  inv.lgaps$Oki_Osa)
# inv.lgaps.Seq$Oki_Kum <- BSgenome::getSeq(OKI2018_I69,  inv.lgaps$Oki_Kum)
# inv.lgaps.Seq$Osa_Oki <- BSgenome::getSeq(OSKA2016v1.9, inv.lgaps$Osa_Oki)
# inv.lgaps.Seq$Osa_Aom <- BSgenome::getSeq(OSKA2016v1.9, inv.lgaps$Osa_Aom)
# inv.lgaps.Seq$Oki_Bar <- BSgenome::getSeq(OKI2018_I69,  inv.lgaps$Oki_Bar)
# inv.lgaps.Seq$Bar_Oki <- BSgenome::getSeq(Bar2_p4,      inv.lgaps$Bar_Oki)
```


## Tandem repeats

Repeated regions of genomes are notoriously difficult to assemble and align. Thus, information about an alignment stop's proximity to a tandem repeat is useful information in breakpoint classification. Given a user-defined tolerance, the function `tan_bp` will classify alignment stops as either far from, close to, or within tandem repeats (elements 1,2 and 3 of the produced list, respectively). The function requires the tandem repeat coverage of one of the genomes. By default, the function will expect the target genome tandem coverage. Setting `query_tf = TRUE` will switch the outcome to be query genome oriented.

```{r echo=T, message=F, warning=F}
# 
# tan_O <- rtracklayer::import.gff3(system.file("extdata/OSKA2016.tantanf4w100.gff3.gz",  package="GenomicBreaks"))
# tan_O <- GRanges(tan_O, seqinfo = seqinfo(OSKA2016))
# tan_gr_O <- tan_bp(gr_ob = gr_O_Oki, tan = tan_O, tol = 50)
# tan_gr_O
# 
# tan_Oki <- rtracklayer::import.gff3(system.file("extdata/I69-5.tantanf4w100.gff3.gz",  package="GenomicBreaks"))
# tan_Oki <- GRanges(tan_Oki, seqinfo = seqinfo(OKI2018_I69))
# tan_gr_Oki <- tan_bp(gr_ob = gr_O_Oki, tan = tan_Oki, tol = 50, query_tf = TRUE)
# tan_gr_Oki

```

We can use another script to plot the density of tandem repeats around alignment stops (and breakpoints). In the following plots, breakpoints are aligned at the centre of the plots. Intuitively, this function can be generalised to see the density of any binary coverage characteristic around breakpoints, given a GRanges object with per-base coverage of a given characteristic. Here we stick with tandem repeat info:

```{r echo=T, message=F, warning=F}
# hm_tan_O <- gr_O_Oki %>%
#   feature_coverage(tan_O, win = 1000, lab = "Osaka ~ Okinawa", direction = "left") %>%
#   smoothHeatmap(output.size = c(2000,500), algorithm = "kernel") %T>%
#   plotHeatmap
# 
# hm_red_tan_O <- reduced_gr_O_Oki %>%
#   feature_coverage(tan_O, win = 1000, lab = "Osaka ~ Okinawa", direction = "left") %>%
#   smoothHeatmap(output.size = c(2000,500), algorithm = "kernel") %T>%
#   plotHeatmap
# 
# hm_tan_Oki <- gr_Oki_O %>%
#   feature_coverage(tan_Oki, win = 1000, lab = "Osaka ~ Okinawa", direction = "left") %>%
#   smoothHeatmap(output.size = c(2000,500), algorithm = "kernel") %T>%
#   plotHeatmapList
```

## Coverage

Another characteristic of alignment stops that we can examine is the coverage depth over said alignment stops. By aligning raw reads to assembled genomes, we can obtain per-base coverage depth for target or query genome. Low coverage directly on and surrounding an alignment stop may suggest unreliability in it being a true breakpoint. The function `bp_coverage` returns a GRanges object of the bps, with associated averaged and point converage in the metadatacolumns. 

```{r echo=T, message=F, warning=F}

# cov_gr_O_Oki <- bp_coverage(gr_ob = q_Oki, cov_gr = Oki_cov_pb, win = 50)
# cov_gr_O_Oki

```

# Examples

## Nucleaic acid content heatmaps

The _GenomicBreaks_ function `bp_heatmap` plots nucleic acid content in windows
centrered on alignment stops.  Lets look at `GC` content before and after
coalescing: 

```{r echo=T, message=F, warning=F}
# library(heatmaps)
# hm_GC1 <- sapply(gbs, bp_heatmap, basep_range = 1000, pat = "GC", direction = "left") |> SimpleList()
# hm_GC2 <- sapply(coa, bp_heatmap, basep_range = 1000, pat = "GC", direction = "left") |> SimpleList()
# 
# maxHeight <- 2000
# sm_GC1 <- sapply(hm_GC1, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# sm_GC2 <- sapply(hm_GC2, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# sm_GC3 <- sapply(hm_GC3, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# 
# plotHeatmapList(sm_GC1)
# plotHeatmapList(sm_GC2)
# plotHeatmapList(sm_GC3)
```

The alignment stops are ordered in such a way that start (or left breaks) are
centred on the top half of the plot, and end alignment stops are centred on the
bottom. This is why we see different directionality in the plots. At the
alignment stops, the GC content seems to be lower, but higher once within the
aligned region. 

Now looking at TATA content:

```{r TATA}
# hm_GC1_TATA <- sapply(gbs, bp_heatmap, basep_range = 1000, pat = "TATA", direction = "left") |> SimpleList()
# hm_GC2_TATA <- sapply(coa, bp_heatmap, basep_range = 1000, pat = "TATA", direction = "left") |> SimpleList()
# 
# maxHeight <- 2000
# sm_GC1_TATA <- sapply(hm_GC1_TATA, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# sm_GC2_TATA <- sapply(hm_GC2_TATA, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# sm_GC3_TATA <- sapply(hm_GC3_TATA, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# 
# plotHeatmapList(sm_GC1_TATA)
# plotHeatmapList(sm_GC2_TATA)
# plotHeatmapList(sm_GC3_TATA)
```

The directionaility here is consistent with the `GC` content analysis.
Furthermore; "TATA box sequence can act as a basal promoter element not only for
RNA polymerase II (RNAP II) transcription, but also for transcription by RNA
polymerase III (RNAP III)" - Wang Y, Jensen RC, Stumph WE. Role of TATA box
sequence and orientation in determining RNA polymerase II/III transcription
specificity. Nucleic Acids Res. 1996;24(15):3100–3106.
doi:10.1093/nar/24.15.3100. However, the areas are not particularly enriched
for TATA boxes, so it does not necessarily show that breakpoints are occurring
directly after promoter regions. 


```{r AAGCsGCwwmkCGrCTTyn}
# hm_GC1_mot <- sapply(gbs, bp_heatmap, basep_range = 1000, pat = "AGCNGC", direction = "left") |> SimpleList()
# hm_GC2_mot <- sapply(coa, bp_heatmap, basep_range = 1000, pat = "AGCNGC", direction = "left") |> SimpleList()
# 
# maxHeight <- 2000
# sm_GC1_mot <- sapply(hm_GC1_mot, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# sm_GC2_mot <- sapply(hm_GC2_mot, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# sm_GC3_mot <- sapply(hm_GC3_mot, smoothHeatmap, output.size = c(maxHeight,     500), algorithm = "kernel") |> SimpleList()
# 
# plotHeatmapList(sm_GC1_mot)
# plotHeatmapList(sm_GC2_mot)
# plotHeatmapList(sm_GC3_mot)
```


## Cluster analysis


```{r cluster_analysis, fig.height=20, fig.width=20}
# # Accessory functions
# clusterHeatmap <- function(hm, k = 2) kmeans(image(hm), k)
# 
# orderHeatmap <- function(hm, cl) {
#   image(hm) <- image(hm)[order(cl$cluster),]
#   hm
# }
# 
# subsetHeatmap <- function(hm, bool) {
#   image(hm) <- image(hm)[bool,]
#   hm
# }
# 
# # DAta
# hm_GC   <- bp_heatmap(      gr_Oki_O, basep_range  = 1000, pat = "GC",     direction = "left")
# hm_TATA <- bp_heatmap(      gr_Oki_O, basep_range  = 1000, pat = "TATA",   direction = "left")
# hm_tan  <- feature_coverage(gr_Oki_O, tan_Oki, win = 1000, lab = "Tandem", direction = "left")
# 
# # Define 3 K-mean clusters
# clusterHeatmap(hm_GC, 3) -> hm_cl_k4
# 
# # Order and smooth
# shm_GC <- orderHeatmap(hm_GC, hm_cl_k4) %>%
#   smoothHeatmap(output.size = c(2000,500), algorithm = "kernel")
# 
# shm_TATA <- orderHeatmap(hm_TATA, hm_cl_k4) %>%
#   smoothHeatmap(output.size = c(2000,500), algorithm = "kernel")
# 
# shm_tan <- orderHeatmap(hm_tan, hm_cl_k4) %>%
#   smoothHeatmap(output.size = c(2000,500), algorithm = "kernel")
# 
# # Plot the ordered data
# plotHeatmapList(list(shm_GC, shm_TATA, shm_tan))
# par(mfrow = c(1,1)) # plotHeatmapList messes with mrfrow...
# plotHeatmapMeta(list(hm_GC, hm_TATA, hm_tan))
# 
# plotHeatmapMeta(list(subsetHeatmap(hm_GC, hm_cl_k4$cluster == 1),
#                      subsetHeatmap(hm_GC, hm_cl_k4$cluster == 2),
#                      subsetHeatmap(hm_GC, hm_cl_k4$cluster == 3)))
# 
# kmeanOrderedPlot <- function (hm, k=3, seed = 1) {
#   set.seed(seed)
#   
#   cl <- kmeans(image(hm), k)
#   mat <- image(hm)[order(cl$cluster),]
#   
#   kmhm = Heatmap(
#     mat,
#     coords=c(-500,500),
#     label="kmeans",
#     scale=range(mat))
#   
#   plotHeatmapList(kmhm,
#                   cex.label=1.5,
#                   partition=as.vector(table(cl$cluster)),
#                   partition.legend=TRUE,
#                   partition.lines=TRUE,
#                   legend=TRUE,
#                   legend.pos="r",
#                   legend.width=0.3)
# }
# 
# kmeanOrderedPlot(shm_GC)
```

## Evidence for breakpoint hotspots

The function `bp_pair_analysis` takes two pairwise alignments with the same target genome, and plots the alignment stops on to the centred alignment stops of the other. In the plot below, Osaka is there consistent target genome, and as such acts as a coordinate system to relate alignment stops across genomes. Lets produce two of these 3-way analyses; one between Osaka-Okinawa-Norway, and one between Osaka-Okinawa-Aomori

```{r echo=T, message=F, warning=F}

# hm1 <- bp_pair_analysis(gr_ref_q1 = gr_O_Oki, gr_ref_q2 = gr_O_N, win = 1000, lab = "Oki~Nor")
# plotHeatmapMeta(hm1)
# 
# hm2 <- bp_pair_analysis(gr_ref_q1 = gr_O_Oki, gr_ref_q2 = gr_O_A, win = 1000, lab = "Oki~Aom")
# plotHeatmapMeta(hm2)

```

The accumulation of alignment breaks of one pairwise alignment onto another suggests the existence of breakpoint hotspots; regions where breaks are far more likely to occur. Intuitively, this would lead to the presumption of synteny blocks, also. The pattern is consistent after coalescing, too;

```{r echo=T, message=F, warning=F}

# reduced_gr_O_N <- coalesce_contigs(gr_ob = gr_O_N, tol = 500)
# 
# hm3 <- bp_pair_analysis(gr_ref_q1 = reduced_gr_O_Oki, gr_ref_q2 = reduced_gr_O_N, win = 1000, lab = "Oki~Nor")
# plotHeatmapMeta(hm3)

```

## Tandem repeat coverage around alignment stops (breakpoints)

The function `feature_coverage` will plot the coverage of tandem repeats around
centred alignment stops. 

```{r echo=T, message=F, warning=F}
# 
# hm_tan1 <- feature_coverage(gr_ob = gr_O_Oki, feat = tan_O, win = 4000, lab = "Osk ~ Oki", direction = "left")
# plotHeatmapMeta(hm_tan1)
#  
# hm_tan2 <- feature_coverage(gr_ob = reduced_gr_O_Oki, feat = tan_O, win = 4000, lab = "OSK~Oki (coalesced)", direction = "left")
# plotHeatmapMeta(hm_tan2)
# 
# # should I also add comparison of proportions of tandem proximities before and after coalescion?
```

There appears to be less tandem repeats around alignment stops after coalescion,
inferring that perhaps that some are artefacts of difficult alignment near to
tandem repeats. 


## Gene feature coverage around alignment stops (breakpoints)

The function `feature_coverage` will plot the coverage of tandem repeats around
centred alignment stops. 

```{r genomic_features}
# #current_GB <- gr_Oki_O # To ease recomputing with another object
# current_GB <- reduced_gr_Oki_O
# hmList_OKI <- list()
# 
# hmList_OKI$genes      <- feature_coverage(current_GB, genes(tx_OKI),                         win = 2000, lab = "Oki genes", direction = "left")
# hmList_OKI$promoters  <- feature_coverage(current_GB, promoters(tx_OKI),                     win = 2000, lab = "Oki promoters", direction = "left")
# hmList_OKI$exons      <- feature_coverage(current_GB, exonicParts(tx_OKI),                   win = 2000, lab = "Oki exons", direction = "left")
# hmList_OKI$introns    <- feature_coverage(current_GB, intronicParts(tx_OKI),                 win = 2000, lab = "Oki introns", direction = "left")
# hmList_OKI$cds        <- feature_coverage(current_GB, cds(tx_OKI),                           win = 2000, lab = "Oki cds", direction = "left")
# hmList_OKI$utrs5      <- feature_coverage(current_GB, unlist(fiveUTRsByTranscript(tx_OKI)),  win = 2000, lab = "Oki 5′ UTRs", direction = "left")
# hmList_OKI$utrs3      <- feature_coverage(current_GB, unlist(threeUTRsByTranscript(tx_OKI)), win = 2000, lab = "Oki 3′ UTRs", direction = "left")
# hmList_OKI$pwm        <- feature_coverage(current_GB, pwmHits                              , win = 2000, lab = "PWM", direction = "left")
# 
# plotHeatmapMeta(hmList_OKI)
# plotHeatmapMeta(hmList_OKI$genes)
# plotHeatmapMeta(hmList_OKI[c("exons", "cds")])
# plotHeatmapMeta(hmList_OKI[c("exons", "introns", "utrs5", "utrs3", "pwm")])
# plotHeatmapMeta(hmList_OKI[c("introns", "utrs5", "utrs3", "pwm")])
# plotHeatmapMeta(hmList_OKI[c("utrs5", "utrs3", "pwm")])
```

Aligned regions are enriched in exons and depleted in introns.  Is it because
exons are more alignable or is it because it is deleterous to break genes
in exons ?

## Gene feature coverage around PWM hits

Let's look at coverage around PWM hits of the AAGCsGCwwmkCGrCTTyn motif

```{r genomic_features_pwmHits}
# current_GB <- pwmHits
# 
# hmList_PWM_OKI <- list()
# hmList_PWM_OKI$genes     <- feature_coverage(current_GB, genes(tx_OKI),                         win = 2000, lab = "Oki genes", direction = "left")
# hmList_PWM_OKI$promoters <- feature_coverage(current_GB, promoters(tx_OKI),                     win = 2000, lab = "Oki promoters", direction = "left")
# hmList_PWM_OKI$exons     <- feature_coverage(current_GB, exonicParts(tx_OKI),                   win = 2000, lab = "Oki exons", direction = "left")
# hmList_PWM_OKI$introns   <- feature_coverage(current_GB, intronicParts(tx_OKI),                 win = 2000, lab = "Oki introns", direction = "left")
# hmList_PWM_OKI$cds       <- feature_coverage(current_GB, cds(tx_OKI),                           win = 2000, lab = "Oki cds", direction = "left")
# hmList_PWM_OKI$utrs5     <- feature_coverage(current_GB, unlist(fiveUTRsByTranscript(tx_OKI)),  win = 2000, lab = "Oki 5′ UTRs", direction = "left")
# hmList_PWM_OKI$utrs3     <- feature_coverage(current_GB, unlist(threeUTRsByTranscript(tx_OKI)), win = 2000, lab = "Oki 3′ UTRs", direction = "left")
# 
# plotHeatmapMeta(hmList_PWM_OKI)
# plotHeatmapMeta(hmList_PWM_OKI[c("exons", "introns")])
# plotHeatmapMeta(hmList_PWM_OKI[c("exons", "utrs5", "utrs3")])
# plotHeatmapMeta(hmList_PWM_OKI[c("utrs5", "utrs3")])
```

They are enriched in introns...

```{r genomic_features3}
# 
# # Sandbox to make sanity checks.
# 
# a <- reduced_gr_Oki_O
# b <- unlist(fiveUTRsByTranscript(tx_OKI))
# 
# hm_a <- feature_coverage(a, b, win = 2000, lab = "All ranges", direction = "left")
# hm_b <- feature_coverage(subsetByOverlaps(a,  b), b, win = 2000, lab = "Subsetted ranges", direction = "left")
# 
# plotHeatmapMeta(list(hm_a))
# plotHeatmapMeta(list(hm_b))
```
