---
title: "Structural Variants"
author: 
 - "Charles Plessy"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Structural Variants}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = FALSE, echo = TRUE)
knitr::opts_knit$set(verbose = TRUE)
library("GenomicBreaks")
```

# Structural Variants

The `GBreaks()` objects here are provided as _lazy data_ and by convention
their name always start with `example`.

## Colinearity

### Definition

We define _structural variants_ as patterns of interruption of the colinearity
between genomes.

Colinearity is a property of pairs of mapped regions.  

```{r colinearRegions_pair}
plotApairOfChrs(exampleColinear, main = "Two colinear regions")
```

Regions that mapped on reverse strands **and** in reverse order are also colinear.

```{r colinearRegions_pair_minus_strand}
plotApairOfChrs(reverse(exampleColinear), main = "Two colinear regions on different strands")
```

Regions that are only in reverse order, or only mapped on opposite strands
**are not colinear**.

```{r notColinearRegions_pair}
plotApairOfChrs(exampleNotColinear, main = "Same strand but not colinear")
plotApairOfChrs(reverse(exampleNotColinear), main = "Both on opposite strand but not colinear")
```

In this example, the syntenic chromosomes are called `chrA` and `chrB`, but the
names do not matter.

### Technical details

Remember that the `flagColinearAlignments()` function puts the flag on the first
member of the pair, ordered by the sequence on the _target_ genome.

```{r colinearRegions_flags}
flagColinearAlignments(exampleColinear)
flagColinearAlignments(reverse(exampleColinear))

# Note the tolerance window
flagColinearAlignments(exampleColinear, tol = 100)

# Strandless mappings can also be colinear
exampleColinear |> plyranges::mutate(strand = "*") |> flagColinearAlignments()
flagColinearAlignments(exampleNotColinear)
```

To make some sanity checks against the other flagging functions there is also
an object containing three colinear mappings.

```{r colinearRegions_triple}
flagColinearAlignments(exampleColinear3)
```

### Limitation

At the moment we can not study the structural variants that are entirely
contained in an aligned region, for instance an indel represented as an
alignment gap.

## Inversions

### Trivial inversions

Trivial inversions are triplets of mapped regions that would be colinear if the
strand of the second one were reversed.  Note that, like for colinear patterns,
the flag is put on the first mapping of the triplet.

```{r inversions}
exampleInversion |> flagColinearAlignments() |> flagInversions()
plotApairOfChrs(exampleInversion, main = "An inverted region flanked by colinear regions")
plotApairOfChrs(reverse(exampleInversion), main = "An inverted region flanked by colinear regions (opposite strands)")

# Removing the inversion restores colinearity
exampleInversion[-2] |> flagColinearAlignments()

# Flipping the inversion restores colinearity
exampleInversion |> plyranges::mutate(strand = "+") |> flagColinearAlignments()

# Note the tolerance window
flagInversions(exampleInversion, tol = 10)
```

### Double inversions

Double inversions are the result of two overlapping trivial inversions, and are
detected with a different approach, implemented in the `flagDoubleInversions()`
function.

Here we construct a double inversion in 3 steps:

 1) First, we start from 5 collinear ranges,
 2) then we invert ranges 2 and 3,
 3) then we invert ranges 3 an 4 (of step 2).
 
 show a triplet resulting from a double inversion, flanked by two
regions that did not move.

```{r doubleInversions}
exampleColinear5 |> plotApairOfChrs(main = "5 collinear ranges")
exampleInversion5uncollapsed |> plotApairOfChrs(main = "Invert ranges 2 and 3")
exampleDoubleInversion1 |> plotApairOfChrs(main = "Invert ranges 3 and 4 of the previous step")
```

Same story but showing coalesced alignments.

```{r doubleInversions2}
exampleColinear5 |> coalesce_contigs() |> plotApairOfChrs()
exampleInversion5uncollapsed |> coalesce_contigs() |> plotApairOfChrs()
exampleDoubleInversion1 |> coalesce_contigs() |> plotApairOfChrs()
```

Given the double inversion example above, but not its history, we can deduce
the coordinates of the inversions (pairs of breakpoints), but not the order
in which they took place.

```{r doubleInversions2}


```

### Nested inversions

The `?flagInversions` algorithm only detects the simplest cases.  In order to
detect nested inversions, the inner ones have to be removed.

```{r nestedInversions}
exampleNestedInversions |> plotApairOfChrs()
exampleNestedInversions |> flagInversions()
(flipped <- (exampleNestedInversions |> flagInversions() |> flipInversions()))
flipped |> plotApairOfChrs()
flipped |> coalesce_contigs() |> flagInversions()
```

## Transolocations

If a region has moved, but is not an inversion, then it is a translocation.

```{r translocation}
exampleTranslocation |> flagColinearAlignments() |> flagInversions()
plotApairOfChrs(exampleTranslocation)
```

It does not matter if the region moved on the minus strand
```{r translocationOnMinus}
exampleTranslocation2 |> flagColinearAlignments() |> flagInversions()
plotApairOfChrs(exampleTranslocation2)
```

## Insertions / Deletions (indels)

NOTE that searching for insertion/deletions make little sense in one-to-one
alignment unless the maximal distance for colinearity is set to a meaningful
value.

```{r indels}
exampleDeletion
plotApairOfChrs(exampleDeletion)
exampleInsertion
plotApairOfChrs(exampleInsertion)
```
