---
title: "Structural Variants"
author: 
 - "Charles Plessy"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Structural Variants}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = FALSE, echo = TRUE)
knitr::opts_knit$set(verbose = TRUE)
library("GenomicBreaks")
```

# Structural Variants

The `GBreaks()` objects here are provided as _lazy data_ and by convention
their name always start with `example`.

## Colinearity

### Definition

We define _structural variants_ as patterns of interruption of the colinearity
between genomes.

Colinearity is a property of pairs of mapped regions.  

```{r colinearRegions_pair}
plotApairOfChrs(exampleColinear, main = "Two colinear regions")
```

Regions that mapped on reverse strands **and** in reverse order are also colinear.

```{r colinearRegions_pair_minus_strand}
plotApairOfChrs(reverse(exampleColinear), main = "Two colinear regions on different strands")
```

Regions that are only in reverse order, or only mapped on opposite strands
**are not colinear**.

```{r notColinearRegions_pair}
plotApairOfChrs(exampleNotColinear, main = "Same strand but not colinear")
plotApairOfChrs(reverse(exampleNotColinear), main = "Both on opposite strand but not colinear")
```

In this example, the syntenic chromosomes are called `chrA` and `chrB`, but the
names do not matter.

### Technical details

Remember that the `flagColinearAlignments()` function puts the flag on the first
member of the pair, ordered by the sequence on the _target_ genome.

```{r colinearRegions_flags}
flagColinearAlignments(exampleColinear)
flagColinearAlignments(reverse(exampleColinear))
# Strandless mappings can also be colinear
exampleColinear |> plyranges::mutate(strand = "*") |> flagColinearAlignments()
flagColinearAlignments(exampleNotColinear)
```

To make some sanity checks against the other flagging functions there is also
an object containing three colinear mappings.

```{r colinearRegions_triple}
flagColinearAlignments(exampleColinear3)
```

### Limitation

At the moment we can not study the structural variants that are entirely
contained in an aligned region, for instance an indel represented as an
alignment gap.

## Inversions

Inversions are triplets of mapped regions that would be colinear if the strand
of the second one were reversed.  Note that, like for colinear patterns, the
flag is put on the first mapping of the triplet.

```{r inversions}
exampleInversion |> flagColinearAlignments() |> flagInversions()
plotApairOfChrs(exampleInversion, main = "An inverted region flanked by colinear regions")
plotApairOfChrs(reverse(exampleInversion), main = "An inverted region flanked by colinear regions (opposite strands)")

# Removing the inversion restores colinearity
exampleInversion[-2] |> flagColinearAlignments()

# Flipping the inversion restores colinearity
exampleInversion |> plyranges::mutate(strand = "+") |> flagColinearAlignments()

# Note the tolerance window
flagInversions(exampleInversion, tol = 10)
```

## Insertions / Deletions (indels)

NOTE that searching for insertion/deletions make little sense in one-to-one
alignment unless the maximal distance for colinearity is set to a meaningful
value.

```{r interruptions}
exampleInsertion |> flagInsersions()
```
