---
title: "Structural Variants"
author: 
 - "Charles Plessy"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Structural Variants}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE, echo = TRUE)
knitr::opts_knit$set(verbose = TRUE)
library("GenomicBreaks")
```

# Structural Variants

The `GBreaks()` objects here are provided as _lazy data_ and by convention
their name always start with `example`.

## Colinearity

We define _structural variants_ as patterns of interruption of the colinearity
between genomes.

At the moment we can not study the structural variants that are entirely
contained in an aligned region, for instance an indel represented as an
alignment gap.

Colinearity is a property of pairs of mapped regions.  Remember that the
`flagColinearAlignments()` function puts the flag on the first member of the
pair, ordered by the sequence on the _target_ genome.

In this example, the syntenic chromosomes are called `chrA` and `chrB`, but the
names do not matter.

```{r colinearRegions_pair}
flagColinearAlignments(exampleColinear)
plotApairOfChrs(exampleColinear)
flagColinearAlignments(reverse(exampleColinear))

# Strandless mappings can also be colinear
exampleColinear |> plyranges::mutate(strand = "*") |> flagColinearAlignments()

flagColinearAlignments(exampleNotColinear)
plotApairOfChrs(exampleNotColinear)
```

To make some sanity checks against the other flagging functions there is also
an object containing three colinear mappings.

```{r colinearRegions_triple}
flagColinearAlignments(exampleColinear3)
plotApairOfChrs(exampleColinear3)
```

## Inversions

Inversions are triplets of mapped regions that would be colinear if the strand
of the second one were reversed.  Note that, like for colinear patterns, the
flag is put on the first mapping of the triplet.

```{r inversions}
exampleInversion |> flagColinearAlignments() |> flagInversions()
plotApairOfChrs(exampleInversion)
exampleInversion[-2] |> flagColinearAlignments()
exampleInversion |> plyranges::mutate(strand = "+") |> flagColinearAlignments()

# Now let's flip the strand...
exampleInversion |> plyranges::mutate(strand = c("+", "+", "+")) |>
  flagColinearAlignments() |> flagInversions()

# Note the tolerance window
flagInversions(flagInversions(exampleInversion - 40, tol = 10))
plotApairOfChrs(exampleInversion - 40)
```

## Insertions / Deletions (indels)

NOTE that searching for insertion/deletions make little sense in one-to-one
alignment unless the maximal distance for colinearity is set to a meaningful
value.

```{r interruptions}
exampleInsertion |> flagInsersions()
```
