---
title: "Aggregate Results"
author: "Charles Plessy"
date: "10/02/2023"
output: 
  html_document: 
    toc: yes
params:
  resultsDir: "path/to/the/yaml/files/"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("ggplot2") |> suppressPackageStartupMessages()
```

How to use this Rmarkdown template in a Singularity image:

`R_LIBS_USER` must be set to an empty value, otherwise `R` will attempt to load
packages from the user's library, which were compiled in a different environment
than the GenomicBreaks Singularity image.

Example command:

```
R_LIBS_USER='' R -e 'rmarkdown::render("inst/rmarkdown/templates/aggregateResults/skeleton/skeleton.Rmd", output_file = "./toto.html", params = list(resultsDir = "/path/do/dir/with/yaml/files"))'
```

# Data load

```{r load_data}
library("yaml")
yamlFiles <- list.files(params$resultsDir, pattern = "*.yaml", full.names = TRUE)
names(yamlFiles) <- yamlFiles |> basename() |> sub(pat = ".yaml", rep="")

getStats <- function(file) {
  y <- yaml::read_yaml(file) |> yaml::yaml.load()
  unlist(y)
}

# getStats_long <- function(file) {
#   pairName <- file |> basename() |> sub(pat = ".yaml", rep="")
#   y <- yaml::read_yaml(file) |> yaml::yaml.load()
#   rbind(
#     list(name = pairName, stat = "unaligned_mean_width", value = y$unaligned_width_summary$Mean),
#     list(name = pairName, stat = "aligned_mean_width", value = y$aligned_width_summary$Mean),
#     list(name = pairName, stat = "similarity", value = 100 - y$alignment_summary$percent_mismatch)
#   )
# }

df <- do.call(rbind, lapply(yamlFiles, getStats)) |> as.data.frame()
df <- df[,colSums(df, na.rm = TRUE) !=0] # Some ranges have always a minimum width of zero, and `cor()` does not like it.
df$species1 <- strsplit(rownames(df), "__") |> lapply(\(.) .[1]) |> unlist()
df$species2 <- strsplit(rownames(df), "__") |> lapply(\(.) .[2]) |> unlist()

# Remove self-comparisons
df <- df[df$species1 != df$species2,]
```

# Results

## Identity and mismatches

The proportion of matches in the alignments is the percentage identity.  We call
the proportion of mismatches in the alignments is the _percentage mismatch_.

Both are computed on the aligned regions, which are shorter and shorter with
evolutionary distance.  Thus, at higher evolutionary distance, the numbers are
computed on a smaller fraction of the genome.  Therefore, even if somewhat
correlate with evolutionary distance, their relationship is not linear.

We will use percent identity or percent mismatch on a case-by-case basis
depending on how well they help to spred another variable on a pairwise plot.

```{r percent_identity_and_mismatch}
df$percent_identity_global   <- df$matches_number_Total    / df$aligned_length_Total * 100
df$percent_mismatches_global <- df$mismatches_number_Total / df$aligned_length_Total * 100
ggplot(df) + geom_point() + aes(percent_identity_global,  percent_mismatches_global)
```

```{r}
makeMatrix <- function(df, column, defaultDiagonal = 100, defaultValue = NA) {
  species <- unique(df$species2)
  m <- matrix(defaultValue, nrow=length(species), ncol=length(species))
  colnames(m) <- rownames(m) <- species
  for (i in 1:length(species)) {
    m[i,i] <- defaultDiagonal
  }
  for (i in 1:nrow(df)) {
    s1 <- df[i, "species1"]
    s2 <- df[i, "species2"]
    if(s1 %in% species)
      m[s1, s2] <- df[i, column]
  }
  m
}
# makeMatrix(df, "customGroup", NA) 
m <- makeMatrix(df, "percent_identity_global")
pheatmap::pheatmap(as.matrix(cluster::daisy(m)))
m <- makeMatrix(df, "percent_identity_global", 100, 50)
pheatmap::pheatmap(as.matrix(m), sym=T)

m <- makeMatrix(df, "percent_mismatches_global", 0, 50)
pheatmap::pheatmap(100 - as.matrix(m), sym=T)
```

## Fraction of the genome aligned

The fraction of the genome aligned appears to negatively correlate with the
proportion of mismatches in aligned regions.  On the other hand, with a percent
identity lower than 65Â %, there is very little correlation at all with the
fraction of the genome aligned.

```{r fraction_genome_aligned}
df$fraction_genome_aligned_target <- df$aligned_target_Total / df$guessed_target_length * 100
df$fraction_genome_aligned_query  <- df$aligned_query_Total  / df$guessed_query_length  * 100
df$fraction_genome_aligned_avg    <- (df$fraction_genome_aligned_target + df$fraction_genome_aligned_query) / 2

ggplot(df) + geom_point() + aes(percent_mismatches_global, fraction_genome_aligned_avg, col = percent_identity_global)
ggplot(df) + geom_point() + aes(percent_identity_global,   fraction_genome_aligned_avg, col = percent_mismatches_global)

ggplot(df) + geom_point() + aes(percent_identity_global,   fraction_genome_aligned_avg) + theme_bw() +
  scale_x_continuous("Percent identity") +
  scale_y_continuous("Fraction of the genome aligned")

```

## Fraction of the genome chained


```{r fraction_genome_chained}
df$fraction_genome_chained_target <- df$chain_target_Total / df$guessed_target_length * 100
df$fraction_genome_chained_query  <- df$chain_query_Total  / df$guessed_query_length  * 100
df$fraction_genome_chained_avg    <- (df$fraction_genome_chained_target + df$fraction_genome_chained_query) / 2

ggplot(df) + geom_point() + aes(percent_mismatches_global, fraction_genome_chained_avg, col = percent_identity_global)
```

## Width of the aligned regions

Length of aligned regions quickly decreases with evolutionary distance and then
sorts of plateaus.

```{r aligned_width}
ggplot(df) + geom_point() + aes(percent_identity_global, aligned_length_Mean, col = percent_mismatches_global)
ggplot(df) + geom_point() + aes(percent_mismatches_global, aligned_length_Mean, col = percent_identity_global)
```

Normalising length of aligned regions by genome size may be useful when
comparing species pairs from distant clades.

```{r aligned_width2}
ggplot(df) + geom_point() + aes(percent_identity_global, aligned_target_Mean / guessed_target_length, col = percent_mismatches_global) + scale_y_log10()
ggplot(df) + geom_point() + aes(percent_identity_global, aligned_query_Mean  / guessed_query_length, col = percent_mismatches_global) + scale_y_log10()
ggplot(df) + geom_point() + aes(percent_identity_global, aligned_length_Mean  / (guessed_target_length + guessed_query_length) / 2, col = percent_mismatches_global)  + scale_y_log10()
```

## Length of the chains

```{r chained_width}
ggplot(df) + geom_point() + aes(percent_identity_global, chain_target_Mean, col = percent_mismatches_global) + scale_y_log10()
ggplot(df) + geom_point() + aes(percent_mismatches_global, chain_target_Mean, col = percent_identity_global) + scale_y_log10()

ggplot(df) + geom_point() + aes(percent_identity_global, chain_target_Mean / guessed_target_length, col = percent_mismatches_global) + scale_y_log10()
ggplot(df) + geom_point() + aes(percent_identity_global, chain_query_Mean  / guessed_query_length, col = percent_mismatches_global) + scale_y_log10()
```

## Indices

The GOC and synteny indices correlate, therefore I will not use the GOC index,
which makes little sense on nucleotide alignments anyway.

```{r plot_similarity_vs_correlation_index}
df$index_avg_synteny      <- ( df$index_synteny_target + df$index_synteny_query ) / 2
df$index_avg_correlation  <- ( df$index_correlation_target + df$index_correlation_query ) / 2
df$index_avg_GOCvicinity4 <- ( df$index_GOCvicinity4_target + df$index_GOCvicinity4_query ) / 2
df$index_avg_strandRand   <- ( df$index_strandRand_target + df$index_strandRand_query ) / 2

df[,grepl("index_avg", colnames(df))] |> pairs()
```

### Synteny index

The synteny index measures how little the homologous genomic regions move between chromosomes.

Some of the dots may be artefacts because of residual haplotypes.

```{r plot_similarity_vs_synteny_index}
ggplot(df) + theme_bw() +
  aes(percent_identity_global, index_avg_synteny, col = percent_mismatches_global) +
  xlab("Identity between aligned regions (%)") +
  ylab("Synteny index") +
  geom_point()
```

### Strand randomisation index

See `?GenomicBreaks::strand_randomisation_index`.

There appears to be a correlation between the strand randomisation index and
the alignment distance.

```{r plot_similarity_vs_correlation_index}
ggplot(df) + theme_bw() +
  aes(percent_identity_global, index_avg_strandRand, col = percent_mismatches_global) +
  xlab("Identity between aligned regions (%)") +
  ylab("Strand randomisation index") +
  geom_point() + geom_smooth()

cor(df$percent_identity_global, df$index_avg_strandRand)

ggplot(df) + theme_bw() +
  aes(percent_mismatches_global, index_avg_strandRand, col = percent_identity_global) +
  xlab("Mismatch between aligned regions (%)") +
  ylab("Strand randomisation index") +
  geom_point() + geom_smooth()

cor(df$percent_mismatches_global, df$index_avg_strandRand)

ggplot(df) + theme_bw() +
  aes(index_avg_synteny, index_avg_strandRand, col = percent_identity_global) +
  xlab("Synteny index") +
  ylab("Strand randomisation index") +
  geom_point() + geom_smooth()
```

### Correlation index

The correlation index measures how "off the diagonal" the alignments are

```{r plot_similarity_vs_correlation_index}
ggplot(df) + theme_bw() +
  aes(percent_identity_global, index_avg_correlation, col = species1) +
  xlab("Identity between aligned regions (%)") +
  ylab("Correlation index") +
  geom_point()
```

### Gene order correlation

```{r plot_similarity_vs_correlation_index}
ggplot(df) +
  aes(percent_identity_global, index_avg_GOCvicinity4) +
  aes(color = species1) +
  xlab("Similarity between aligned regions") +
  ylab("abs(correlation index)") +
  geom_point()
```
